docs/dev/howtos/development-with-android-components.md

```diff
@@ -0,0 +1,82 @@
+
+Modern Gradle supports [composite builds](https://docs.gradle.org/current/userguide/composite_builds.html), which allows to substitute on-disk projects for binary publications.  Composite builds transparently accomplish what is usually a frustrating loop of:
+1. change library
+1. publish library snapshot to the local Maven repository
+1. consume library snapshot in application
+
+## Preparation
+
+Clone the Glean SDK and android-components repositories:
+
+```sh
+git clone https://github.com/mozilla/glean
+git clone https://github.com/mozilla-mobile/android-components
+```
+
+## Cargo build targets
+
+By default when building Android Components using gradle, the gradle-cargo plugin will compile Glean for every possible platform,
+which might end up in failures.
+You can customize which targets are built in `glean/local.properties`
+([more information](https://github.com/ncalexan/rust-android-gradle/blob/master/README.md#specifying-local-targets)):
+
+```
+# For physical devices:
+rust.targets=arm
+
+# For unit tests:
+# rust.targets=darwin # Or linux-*, windows-* (* = x86/x64)
+
+# For emulator only:
+rust.targets=x86
+```
+
+## Substituting projects
+
+android-components has custom build logic for dealing with composite builds,
+so you should be able to configure it by simply adding the path to the Glean repo in the correct `local.properties` file:
+
+In `android-components/local.properties`:
+```groovy
+substitutions.glean.dir=../glean
+```
+
+If this doesn't seem to work, or if you need to configure composite builds for a project that does not contain this custom logic,
+add the following to `settings.gradle`:
+
+In `android-components/settings.gradle`:
+```groovy
+includeBuild('../glean') {
+  dependencySubstitution {
+    substitute module('org.mozilla.telemetry:glean') with project(':glean-core')
+    substitute module('org.mozilla.telemetry:glean-forUnitTests') with project(':glean-core')
+  }
+}
+```
+
+## Caveat
+
+There's a big gotcha with library substitutions: the Gradle build computes lazily, and AARs don't include their transitive dependencies' JNI libraries.
+This means that in `android-components`, `./gradlew :service-glean:assembleDebug` **does not** invoke `:glean-core:cargoBuild`,
+even though `:service-glean` depends on the substitution for `:glean-core` and even if the inputs to Cargo have changed!
+It's the final consumer of the `:service-glean` project (or publication) that will incorporate the JNI libraries.
+
+In practice that means _you should always be targeting something that produces an APK_: a test or a sample module.
+Then you should find that the `cargoBuild` tasks are invoked as you expect.
+
+`./gradlew :samples-glean:connectedAndroidTest` should work.  
+Other tests like `:glean-core:testDebugUnitTest` or `:support-sync-telemetry:testDebugUnitTest` will currently fail, because the JNI libraries are not included.
```

Comment:
I find this section a bit confusing since it doesn't specify which project (a-c or glean.rs) that these tests are run in, and it mixes targets in `a-c` (`samples-glean`) and glean (`glean-core`).

I think the general idea here is that you don't have to re-run anything the the glean checkout, just that you have to run a apk-producing target in the `a-c` checkout.  Is that correct?

Reply:
o_O Woops, the mix of samples-glean and glean-core is a copy-pasta mistake. This should all be run inside a-c. I'll fix that and make it more clear it's all in a-c.